pkg/cache/cache.go

```diff
@@ -166,6 +177,9 @@ type Options struct {
 	//
 	// This is a global setting for all objects, and can be overridden by the ByObject setting.
 	UnsafeDisableDeepCopy *bool
+
+	// NewInformer allows overriding of NewSharedIndexInformer for testing.
+	NewInformer func(toolscache.ListerWatcher, runtime.Object, time.Duration, toolscache.Indexers) toolscache.SharedIndexInformer
```

Comment:
Can we remove this so we don't have to extend our public API for tests?

Reply:
To clarify, you'd like to use informertest.FakeInformer in the cache package? 

> Drop the type assertion from the FakeInformer package, removing the need for the import but reducing likelihood of drift detection

Could we move the type assertion in the cache package test?

Reply:
> To clarify, you'd like to use informertest.FakeInformer in the cache package?

Correct. The test currently uses the fake informer to emulate an informer that never syncs:

```golang
	NonBlockingGetTest(cache.MultiNamespacedCacheBuilder([]string{testNamespaceOne, testNamespaceTwo, "default"}), cache.Options{
		NewInformer: func(lw kcache.ListerWatcher, o runtime.Object, d time.Duration, i kcache.Indexers) kcache.SharedIndexInformer {
			return &controllertest.FakeInformer{Synced: false}
		},
	})
```

With the public interface change in this PR, I can write a test like the above but still have a separate test package (`cache_test`). If we were to make this override ability private, the `cache_test` package would become part of `cache`, so it could access the private methods.

> Could we move the type assertion in the cache package test?

The circular import problem comes from the import chain:

```
 "cache_test" => "sigs.k8s.io/controller-runtime/pkg/controller/controllertest"
 "sigs.k8s.io/controller-runtime/pkg/controller/controllertest" => "cache"
```

If `cache_test` becomes `cache`, the import is circular.

I was wrong when I said it was just a type assertion. It looks like controllertest depends heavily on the "cache" package. Example:

https://github.com/kubernetes-sigs/controller-runtime/blob/a23f3603e1dc14e1d14620b814662f55d1987b02/pkg/controller/controllertest/util.go#L86-L94

Reply:
Any feedback as to which approach people would like to see?

Reply:
@alvaroaleman I'll leave this up to you, I'm ok either way as long as we have tests

Reply:
Allright, lets leave it in

Reply:
@sbueringer suggested using reflect in the unit tests. Trying that on. LMK if it LGTY.

Reply:
Linking to reflect comment for reference: https://github.com/kubernetes-sigs/controller-runtime/pull/2371/files/35b504b14c79305b5fee50ea039d514b34c39d62#r1297426919